### PR TITLE
Discount seller liability rule

### PR DESCRIPTION
**Pull request by arshvirk**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,4 @@
 LIGHTRAIL_API_PATH=https://api.lightrail.com/v2
 
 # Lightrail API key
-LIGHTRAIL_API_KEY=
+LIGHTRAIL_API_KEY="eyJ2ZXIiOjMsInZhdiI6MSwiYWxnIjoiSFMyNTYiLCJ0eXAiOiJKV1QifQ.eyJnIjp7Imd1aSI6InVzZXItM2UwZjExMjNlZmY3NDNmYzgwYjRlMWRlMWFkYTk5NTItVEVTVCIsImdtaSI6InVzZXItM2UwZjExMjNlZmY3NDNmYzgwYjRlMWRlMWFkYTk5NTItVEVTVCIsInRtaSI6InVzZXItM2UwZjExMjNlZmY3NDNmYzgwYjRlMWRlMWFkYTk5NTItVEVTVCJ9LCJhdWQiOiJBUElfS0VZIiwiaXNzIjoiU0VSVklDRVNfVjEiLCJpYXQiOjE1NzM3NzE2ODUuNzYyLCJqdGkiOiJiYWRnZS1lNThmNGU2NTBkOWY0ZTcxYTQxODRhN2MwZDFkYmFkNyIsInBhcmVudEp0aSI6ImJhZGdlLTY2OWY2YmM2MmIzZTRmNzliYmY3ZWM5ZTBmM2RhZjIwIiwic2NvcGVzIjpbXSwicm9sZXMiOlsiYWNjb3VudE1hbmFnZXIiLCJjb250YWN0TWFuYWdlciIsImN1c3RvbWVyU2VydmljZU1hbmFnZXIiLCJjdXN0b21lclNlcnZpY2VSZXByZXNlbnRhdGl2ZSIsInBvaW50T2ZTYWxlIiwicHJvZ3JhbU1hbmFnZXIiLCJwcm9tb3RlciIsInJlcG9ydGVyIiwic2VjdXJpdHlNYW5hZ2VyIiwidGVhbUFkbWluIiwid2ViUG9ydGFsIl19.CWjn5qrm3XYg8n5IoD6uXKsOVrtetJW5pxM4nxv59Uw"
```
Comment (Tim-Jordison) [inline .env.example]: This shouldn't have been changed. Even if it's removed, it's not actually great that this production API key will be in the commit history. I think we'll want to find a way to reset this.

Comment (Tim-Jordison) [reply .env.example]: https://stackoverflow.com/questions/18357511/git-remove-committed-file-after-push

Comment (Tim-Jordison) [reply .env.example]: The reason is to why this isn't great is this is a public repo. So your API is publicly accessible while it is in the commit history. As per the link above, it looks like there should be a way to remove this file from that commit.

File: src/model/Program.ts [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 import {RedemptionRule} from "./RedemptionRule";
+import {DiscountSellerLiabilityRule} from "./DiscountSellerLiabilityRule";
 
 export interface Program {
     id: string;
     name: string;
     currency: string;
     discount: boolean;
     discountSellerLiability: number;
+    discountSellerLiabilityRule?: DiscountSellerLiabilityRule;
```
Comment (Tim-Jordison) [inline src/model/Program.ts]: It strikes me as a little surprising that this is the only property that's optional in this interface. How is it different from `balanceRule`?

File: src/model/Value.ts [hunk 3]
```diff
@@ -21,6 +22,7 @@ export interface Value {
     redemptionRule: RedemptionRule;
     discount: boolean;
     discountSellerLiability: number;
+    discountSellerLiabilityRule: DiscountSellerLiabilityRule;
```
Comment (Tim-Jordison) [inline src/model/Value.ts]: Not optional here? I think this is right, but it should be consistent between Programs and Values.

File: src/params/programs/CreateProgramParams.ts [hunk 4]
```diff
@@ -2,13 +2,15 @@ import {Program} from "../../model";
...
 import {RedemptionRule} from "../../model/RedemptionRule";
+import {DiscountSellerLiabilityRule} from "../../model/DiscountSellerLiabilityRule";
 
 export interface CreateProgramParams {
     id: string;
     currency: string;
     name: string;
     discount?: boolean;
     discountSellerLiability?: number;
+    discountSellerLiabilityRule?: DiscountSellerLiabilityRule;
```
Comment (Tim-Jordison) [inline src/params/programs/CreateProgramParams.ts]: Optional here makes sense. 👍

File: src/programs.test.ts [hunk 5]
```diff
@@ -47,9 +50,9 @@ describe("programs", () => {
             };
             const program = await Lightrail.programs.createProgram(request);
             chai.assert.isNotNull(program);
-            chai.assert.deepEqualExcluding(program.body, request,
+            chai.assert.deepEqualExcluding(program.body, {...request, discountSellerLiability: 0.5},
```
Comment (Tim-Jordison) [inline src/programs.test.ts]: Nice work.

File: src/params/programs/UpdateProgramParams.ts [hunk 6]
```diff
@@ -1,10 +1,12 @@
 import {BalanceRule, Program, RedemptionRule} from "../../model";
 import {LightrailResponse} from "../LightrailResponse";
+import {DiscountSellerLiabilityRule} from "../../model/DiscountSellerLiabilityRule";
 
 export interface UpdateProgramParams {
```
Comment (Tim-Jordison) [inline src/params/programs/UpdateProgramParams.ts]: Nice catch on adding `UpdateProgramParams`. I think you're missing `UpdateValueParams` though.

File: src/model/Program.ts (hunk 2 above)
Comment (arshvirk) [reply src/model/Program.ts]: fixed

File: src/model/DiscountSellerLiabilityRule.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export interface DiscountSellerLiabilityRule {
```
Comment (pushplay) [inline src/model/DiscountSellerLiabilityRule.ts]: This should also be exported from src/model/index.ts.

File: src/model/Program.ts [hunk 8]
```diff
@@ -1,12 +1,14 @@
 import {BalanceRule} from "./BalanceRule";
 import {RedemptionRule} from "./RedemptionRule";
+import {DiscountSellerLiabilityRule} from "./DiscountSellerLiabilityRule";
 
 export interface Program {
     id: string;
     name: string;
     currency: string;
     discount: boolean;
     discountSellerLiability: number;
```
Comment (pushplay) [inline src/model/Program.ts]: discountSellerLiability should be marked as `@deprecated` everywhere it's used.

File: src/model/DiscountSellerLiabilityRule.ts (hunk 7 above)
Comment (arshvirk) [reply src/model/DiscountSellerLiabilityRule.ts]: Thanks, i missed that

**Pull request merged**